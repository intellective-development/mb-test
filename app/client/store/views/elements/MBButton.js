// @flow

import * as React from 'react';
import bindClassNames from '../../../shared/utils/bind_classnames';

import styles from './MBButton.scss';

const cn = bindClassNames(styles);

// This module exports a wrapper around the basic html button element.
// The idea is to provide a new baseline for us to work off of as we move towards component style sharing.
// Eventually, the goal is to fit this component and its props into the styleguide the design team is working towards.

type MBButtonProps = {
  className?: string,
  type?: 'default' | 'hollow' | 'action',
  size?: 'medium' | 'small' | 'tall', // TODO: transition to a boolean, small or not small
  expand?: boolean
};

const MBButton = ({className, type = 'default', size = 'medium', expand = false, ...rest_props}: MBButtonProps) => {
  const classes = cn('elMBButton', {
    // types
    elMBButton__Default: type === 'default',
    elMBButton__Hollow: type === 'hollow',
    elMBButton__Action: type === 'action',

    // sizes
    elMBButton__Medium: size === 'medium',
    elMBButton__Small: size === 'small',
    elMBButton__Tall: size === 'tall',

    elMBButton__Expand: expand
  }, className);

  return <button {...rest_props} className={classes} />;
};

export default MBButton;
