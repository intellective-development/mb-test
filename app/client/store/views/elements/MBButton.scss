@import "../style/variables";

// this is still somewhat relying on the global button styles
// prefixed with button to help with some stuff
.elMBButton {
  outline: none; // turn off default system interaction styles
  // from .button
  -webkit-appearance: none;
  border: 1px solid $brand-black;
  border-radius: 4px;
  color: $brand-white;
  cursor: pointer;
  font-family: $font-family-avenir;
  font-weight: $heavy-font-weight;
  line-height: 1.3em;
  margin: 0;
  padding-bottom: 0.75em;
  padding-top: 0.875em;
  text-transform: uppercase;
  transition: background-color 300ms ease-out, border-color 300ms ease-out;

  &:focus {
    color: initial;
    background-color: initial;
  }
}

.elMBButton__Default {
  @include pixel-antialiased;
  @include button-disabled-color(
    $brand-black, // unchanged
    $brand-black, // unchanged
    $brand-white, // unchanged
    0.8
  );

  background-color: $brand-black;
  border-color: $brand-black;
  color: $brand-white;

  &:hover,
  &:active {
    background-color: lighten($brand-black, 20%);
    border-color: lighten($brand-black, 20%);
  }
}

.elMBButton__Hollow {
  @include button-disabled-color(
    $brand-transparent, // unchanged
    desaturate($brand-black, 40%),
    desaturate($brand-black, 40%)
  );
  background-color: $brand-transparent;
  border-color: $brand-black;
  color: $brand-black;

  &:hover,
  &:active {
    background-color: $brand-pale-blue;
    color: $brand-black;
  }
}

.elMBButton__Action {
  @include pixel-antialiased;
  @include button-disabled-color(
    desaturate($brand-red, 70%),
    desaturate($brand-red, 70%),
    $brand-white // unchanged
  );

  background: $brand-red;
  border-color: $brand-red;
  color: $brand-white;

  &:hover,
  &:active {
    background-color: lighten($brand-red, 10%);
    border-color: lighten($brand-red, 10%);
  }
}

// TODO: consolidate medium/tall, they're the default size
.elMBButton__Medium {
  @include font-size(14);

  letter-spacing: 0.125rem;
}

.elMBButton__Tall {
  @include font-size(14);

  letter-spacing: 0.125rem;
  min-height: 48px;
}

.elMBButton__Small {
  @include font-size(13);

  letter-spacing: 0;

  // goal is 32px tall
  min-height: 32px;
  padding: 0.6em 1.125em 0.35em;
  text-transform: capitalize;
}

.elMBButton__Expand {
  width: 100%;
}
