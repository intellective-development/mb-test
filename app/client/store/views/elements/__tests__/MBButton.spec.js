
import * as React from 'react';

// import TestProvider from 'src/views/__tests__/utils/TestProvider';
import MBButton from '../MBButton';

describe('MBButton', () => {
  it('renders hollow buttons', () => {
    expect(shallow(<MBButton type="hollow" />)).toMatchSnapshot();
  });

  it('renders small buttons', () => {
    expect(shallow(<MBButton size="small" />)).toMatchSnapshot();
  });

  it('renders large buttons', () => {
    expect(shallow(<MBButton size="large" />)).toMatchSnapshot();
  });

  it('renders buttons that expand', () => {
    expect(shallow(<MBButton expand />)).toMatchSnapshot();
  });

  it('renders with many props', () => {
    expect(shallow(<MBButton type="hollow" size="small" expand />)).toMatchSnapshot();
  });
});
