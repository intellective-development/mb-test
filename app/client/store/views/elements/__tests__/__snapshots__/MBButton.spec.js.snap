// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MBButton renders buttons that expand 1`] = `
<button
  className="elMBButton elMBButton__Default elMBButton__Medium elMBButton__Expand"
/>
`;

exports[`MBButton renders hollow buttons 1`] = `
<button
  className="elMBButton elMBButton__Hollow elMBButton__Medium"
/>
`;

exports[`MBButton renders large buttons 1`] = `
<button
  className="elMBButton elMBButton__Default"
/>
`;

exports[`MBButton renders small buttons 1`] = `
<button
  className="elMBButton elMBButton__Default elMBButton__Small"
/>
`;

exports[`MBButton renders with many props 1`] = `
<button
  className="elMBButton elMBButton__Hollow elMBButton__Small elMBButton__Expand"
/>
`;
